### PR TITLE
Login/logout react

### PR DESCRIPTION
**Pull request by gon-escalante**
## Summary
Added a login screen working with the books API and dropdowns for the profile and notifications. The profile has a logout button working and a mocked profile button. The notifications dropdown has one mocked notification. Also the WBooks logo now redirects you to the dashboard.

## ScreenShots
peek 2017-12-12 16-22

## Trello Card
https://trello.com/c/8gRpB8HH/9-login-layout

### CODE REVIEW

File: wbooks/src/app/components/Header/index.js [hunk 1]
```diff
@@ -1,26 +1,102 @@
...
 
 
 const Header = () =>(
   <div className="header">
-    <img 
-    className="logo"
-    alt="logo"
-    src={logo}/>
-    <Menu/>
+    <Link to='/'><img
```
Comment (ignaciosantise) [inline wbooks/src/app/components/Header/index.js]: Use the `routes` constant here

File: wbooks/src/app/components/Header/index.js [hunk 2]
```diff
@@ -1,26 +1,102 @@
...
+      className="logo"
+      alt="logo"
+      src={logo}/>
+    </Link>
+    {localStorage.getItem('isLoggedIn') &&
```
Comment (ignaciosantise) [inline wbooks/src/app/components/Header/index.js]: define a file for all localStorage functions, maybe in `utils` or `config` folder

File: wbooks/src/app/components/Header/index.js [hunk 3]
```diff
@@ -1,26 +1,102 @@
...
+    this.setState({showProfileDropdown: false});
+  }
+
+  closeNotificationsDropdown = () => {
+    this.setState({showNotificationsDropdown: false});
+  }
+
+  handleLogout = (e) => {
+    localStorage.removeItem('isLoggedIn');
+    localStorage.removeItem('accessToken');
```
Comment (ignaciosantise) [inline wbooks/src/app/components/Header/index.js]: move these functions to a local storage file

File: wbooks/src/app/components/Header/index.js [hunk 4]
```diff
@@ -1,26 +1,102 @@
...
-  <div className="menu">
-    <img className="menu-item" alt="notification_img" src={notification_img}/>
-    <img className="menu-item" alt="add_book_img" src={add_book_img}/>
-    <img className="menu-item menu-profile-pic" alt="menu-profile-pic" src="https://upload.wikimedia.org/wikipedia/commons/thumb/2/28/Duck_wings_outstretched.jpg/1200px-Duck_wings_outstretched.jpg"/>
-  </div>
-  );
+class Menu extends React.Component{
+  state = {showProfileDropdown: false, isLoggedIn: localStorage.getItem('isLoggedIn')};
+
+  triggerProfileDropdown = (e) => {
```
Comment (ignaciosantise) [inline wbooks/src/app/components/Header/index.js]: if you dont need `e`, remove it from params

File: wbooks/src/app/components/Header/index.js [hunk 5]
```diff
@@ -1,26 +1,102 @@
...
   );
 
-const Menu = () => (
-  <div className="menu">
-    <img className="menu-item" alt="notification_img" src={notification_img}/>
-    <img className="menu-item" alt="add_book_img" src={add_book_img}/>
-    <img className="menu-item menu-profile-pic" alt="menu-profile-pic" src="https://upload.wikimedia.org/wikipedia/commons/thumb/2/28/Duck_wings_outstretched.jpg/1200px-Duck_wings_outstretched.jpg"/>
-  </div>
-  );
+class Menu extends React.Component{
```
Comment (ignaciosantise) [inline wbooks/src/app/components/Header/index.js]: `import React, { Component } from 'react'`;

and

`class Menu extends Component {`

File: wbooks/src/app/components/Header/index.js [hunk 6]
```diff
@@ -1,26 +1,102 @@
...
+  state = {showProfileDropdown: false, isLoggedIn: localStorage.getItem('isLoggedIn')};
+
+  triggerProfileDropdown = (e) => {
+    if(!this.state.showProfileDropdown){
+      this.setState({showProfileDropdown:true});
+      this.profileDropdownMenu.focus();
+    }
+    else
+      this.setState({showProfileDropdown:false});
+  }
```
Comment (ignaciosantise) [inline wbooks/src/app/components/Header/index.js]: you could simplify this using `prevState`

```
this.setState(prevState => {
   if(prevState.showProfileDropdown) {
      this.profileDropdownMenu.focus();
   }
   return { showProfileDropdown: !prevState.showProfileDropdown };
}
```
Read this!
https://reactjs.org/docs/state-and-lifecycle.html#state-updates-may-be-asynchronous

File: wbooks/src/app/components/Header/index.js [hunk 7]
```diff
@@ -1,26 +1,102 @@
...
+
+  triggerNotificationsDropdown = (e) => {
```
Comment (ignaciosantise) [inline wbooks/src/app/components/Header/index.js]: remove `e` from params

File: wbooks/src/app/components/Header/index.js [hunk 8]
```diff
@@ -1,26 +1,102 @@
...
+    if(!this.state.showNotificationsDropdown){
+      this.setState({showNotificationsDropdown:true});
+      this.notificationsDropdownMenu.focus();
+    }
+    else
+      this.setState({showNotificationsDropdown:false});
+  }
```
Comment (ignaciosantise) [inline wbooks/src/app/components/Header/index.js]: same refactor using prevState

File: wbooks/src/app/components/Header/index.js [hunk 9]
```diff
@@ -1,26 +1,102 @@
...
+    localStorage.removeItem('isLoggedIn');
+    localStorage.removeItem('accessToken');
+    this.setState({showProfileDropdown: false, isLoggedIn: false});
+  }
+
+
+  render(){
+    return(
+      <div>
+        { !this.state.isLoggedIn && <Redirect to='/'/> }
```
Comment (ignaciosantise) [inline wbooks/src/app/components/Header/index.js]: do you need this?

File: wbooks/src/app/index.js [hunk 10]
```diff
@@ -17,13 +18,26 @@ class App extends Component {
...
+          <PrivateRoute path={`${BOOKS}/:id`} component={BookDetail}/>
+          <Route path={LOGIN} component={Login}/>
           <Route component={NotFoundPage}/>
         </Switch>
       </div>
     </Router>
     );
   }
 }
+const PrivateRoute = ({ component: Component, ...rest }) => (
```
Comment (ignaciosantise) [inline wbooks/src/app/index.js:32]: 👏

File: wbooks/src/app/screens/bookDetail/index.js [hunk 11]
```diff
@@ -16,6 +16,7 @@ class BookDetail extends Component {
     this.setState({book: this.getBookInfo(parseInt(this.props.match.params.id, 10))});
   }
 
+
```
Comment (ignaciosantise) [inline wbooks/src/app/screens/bookDetail/index.js]: good line! 😆

File: wbooks/src/app/screens/login/index.js [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+      authService.login({mail: this.state.mail, pass: this.state.pass}, this.onLoginSuccess, this.onLoginError);
+    }
+  }
+
+  validateEmail(email) {
+    var re = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
+    return re.test(email);
+  }
+
+  validatePass(pass){
```
Comment (ignaciosantise) [inline wbooks/src/app/screens/login/index.js]: move both validation functions to `utils/validations.js` and import them here

File: wbooks/src/app/screens/login/index.js [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    this.setState({pass: e.target.value});
+  }
+
+  handleMailBlur = (e) => {
+    if(!this.validateEmail(e.target.value))
+      this.setState({mailError: true});
+    else{
+      this.setState({mailError: false});
+    }
+  }
```
Comment (ignaciosantise) [inline wbooks/src/app/screens/login/index.js]: `this.setState({ mailError: !this.validateEmail(e.target.value) });`

File: wbooks/src/app/screens/login/index.js [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+
+  handlePassBlur = (e) => {
+    if(!this.validatePass(e.target.value))
+      this.setState({passLengthError: true});
+    else
+      this.setState({passLengthError: false});
+  }
```
Comment (ignaciosantise) [inline wbooks/src/app/screens/login/index.js]: `this.setState({ passLengthError: !this.validatePass(e.target.value) });`

File: wbooks/src/app/screens/login/index.js [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+
+  render() {
+    const { from } = this.props.location.state || { from: { pathname: ROOT } }
```
Comment (ignaciosantise) [inline wbooks/src/app/screens/login/index.js]: `const from =  this.props.location.state.from || { pathname: ROOT }` ?

File: wbooks/src/app/screens/login/index.js [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+
+    if (redirectToReferrer || localStorage.getItem('isLoggedIn')) {
+      return (
+        <Redirect to={from}/>
+      )
+    }
+
+    return (
+      <div className="login-page">
+        <p className="login-error">You must log in to view the page at {from.pathname}</p>
```
Comment (ignaciosantise) [inline wbooks/src/app/screens/login/index.js]: the app is in spanish 🇪🇸

**Review by ignaciosantise — CHANGES_REQUESTED** [review]

File: wbooks/src/app/screens/bookDetail/index.js (hunk 11 above)
Comment (gon-escalante) [reply wbooks/src/app/screens/bookDetail/index.js]: vital

File: wbooks/src/app/screens/login/index.js (hunk 15 above)
Comment (gon-escalante) [reply wbooks/src/app/screens/login/index.js]: makes sense

**Review by ignaciosantise — APPROVED** [review]

**Pull request merged**